# Курсовая работа № 2
## Аналитика. Начальный уровень

Курсовая работа состоит из двух частей – обязательной и дополнительной. **Для зачета необходимо выполнение только первой части.** Выполнение второй части может потребовать дополнительные знания Python.

- [Часть первая](#Часть-первая)
- [Часть вторая](#Часть-вторая)

## Часть первая


Перед вами стоит задача – подготовить аналитический отчет для HR-отдела. На основании проведенной аналитики предполагается составить рекомендации для отдела кадров по стратегии набора персонала, а также по взаимодействию с уже имеющимися сотрудниками.
<br><br> В базе данных лежит набор таблиц, которые содержат данные о сотрудниках вымышленной компании.
Сделайте обзор штата сотрудников компании. Составьте набор предметов исследования, а затем проверьте их на данных. Вся аналитика должна быть выполена с помощью SQL. Впоследствии данные можно визуализировать, однако финальные датафреймы для графиков также должны быть подготовлены с помощью SQL. <br><br>

Примеры гипотез:
1. Есть зависимость между `perfomance score` и тем, под чьим руководством работает сотрудник.
2. Есть зависимость между продолжительностью работы в компании и семейным положением сотрудника.
2. Есть зависимость между продолжительностью работы в компании и возрастом сотрудника.

<br><br>
Параметры для подключения следующие: хост – `dsstudents.skillbox.ru`, порт – `5432`, имя базы данных – `human_resources`, пользователь – `readonly`, пароль – `6hajV34RTQfmxhS`. Таблицы, доступные для анализа, – `hr_dataset`, `production_staff`, `recruting_costs`, `salary_grid`.

In [1]:
### YOUR CODE HERE ###

import psycopg2
import pandas as pd
import numpy as np
import sqlite3

In [295]:
# пять запросов ниже не нужны для аналитики - только чтобы я мог посмотреть, что делается в sql-таблицах

sql_str = []
columns_list = [[], [], [], [], []]

columns = ('marriedid', 'maritalstatusid', 'genderid', 'empstatus_id', 'deptid', 'state', 'sex', \
           'citizendesc', '"Hispanic/Latino"', 'racedesc', 'department', '"Manager Name"', '"Employee Source"')

sql_str.append("SELECT * FROM hr_dataset;")
sql_str.append("SELECT * FROM production_staff;")
sql_str.append("SELECT * FROM recruiting_costs;")
sql_str.append("SELECT * FROM salary_grid;")
sql_str.append("SELECT * FROM information_schema.columns WHERE table_schema='public';")

# Для аналитики

for i in columns:

    sql_str.append('SELECT ' + i + ', count("Days Employed"), avg("Days Employed") \
                    FROM hr_dataset \
                    WHERE "Date of Termination" is not null GROUP BY ' + i + ';')
    columns_list.append([i, 'count of workers', 'average Days Employed'])

    
    sql_str.append('SELECT que1."Performance", que1.' + i + ', que1."quantity", que2."sum_quantity", CAST(que1."quantity" AS float) / que2."sum_quantity" \
                   FROM (SELECT \
                    CASE \
                       WHEN "Performance Score" IN (\'Exceeds\', \'Exceptional\', \'Fully Meets\') \
                       THEN \'Good performance\' \
                       WHEN "Performance Score" = \'Needs Improvement\' \
                       THEN \'No good performance\' \
                       ELSE \'Others\' \
                   END AS "Performance", \
                   ' + i + ', count(*) AS quantity \
                   FROM hr_dataset \
                   GROUP BY "Performance", ' + i + ') que1 \
                INNER JOIN \
                   (SELECT ' + i +', count(*) as sum_quantity \
                   FROM hr_dataset \
                   GROUP BY ' + i + ') que2 \
               ON que1.' + i + ' = que2.' + i + ' \
               ORDER BY "Performance", ' + i)
    columns_list.append(['Performance', i, 'quantity of workers per group', 'workers in group ' + i, "share of workers"])

# Для проверки тождественности полей
    
sql_str.append('SELECT count(*) FROM \
(SELECT \
    CASE \
        WHEN maritalstatusid = 0 \
        THEN \'Single\' \
        WHEN maritalstatusid = 1 \
        THEN \'Married\' \
        WHEN maritalstatusid = 2 \
        THEN \'Divorced\' \
        WHEN maritalstatusid = 3 \
        THEN \'Separated\' \
        WHEN maritalstatusid = 4 \
        THEN \'Widowed\' \
    END as marital, maritaldesc \
FROM hr_dataset) AS que \
WHERE marital <> maritaldesc')
columns_list.append(['count of maritalstatusid <> maritaldesc'])

sql_str.append('SELECT count(*) FROM \
(SELECT \
    CASE \
        WHEN genderid = 0 \
        THEN \'Female\' \
        WHEN genderid = 1 \
        THEN \'Male\' \
    END as gender, sex \
FROM hr_dataset) AS que \
WHERE gender <> sex')
columns_list.append(['genderid <> sex'])

sql_str.append('SELECT count(*) FROM \
(SELECT \
    CASE \
        WHEN empstatus_id = 1 \
        THEN \'Active\' \
        WHEN empstatus_id = 2 \
        THEN \'Future Start\' \
        WHEN empstatus_id = 3 \
        THEN \'Leave of Absence\' \
        WHEN empstatus_id = 4 \
        THEN \'Terminated for Cause\' \
        WHEN empstatus_id = 5 \
        THEN \'Voluntarily Terminated\' \
    END as empstatus, "Employment Status" \
FROM hr_dataset) AS que \
WHERE empstatus <> "Employment Status"')
columns_list.append(['empstatus_id <> Employment Status'])

In [296]:
# Выполним запросы и заберем данные из таблиц БД

pg_list = []

pg_connection = {
    "host": "dsstudents.skillbox.ru",
    "port": 5432,
    "dbname": "human_resources",
    "user": "readonly",
    "password": "6hajV34RTQfmxhS"
}
conn = psycopg2.connect(**pg_connection)

cursor = conn.cursor()

for i in sql_str:
    cursor.execute(i)
    pg_list.append([a for a in cursor.fetchall()])

conn.commit()

In [188]:
# служебный блок, для аналитики не нужен - только для моего удобства

info_df = pd.DataFrame(pg_list[4])

hr_df = pd.DataFrame(pg_list[0], columns = info_df.loc[info_df[2] == 'hr_dataset'][3].tolist(), \
                    index = np.array(pg_list[0])[:,0])
hr_df = hr_df.drop('id', axis=1)
hr_df.dob = pd.to_datetime(hr_df.dob, errors = 'coerce')
hr_df['Date of Hire'] = pd.to_datetime(hr_df['Date of Hire'], errors = 'coerce')
hr_df['Date of Termination'] = pd.to_datetime(hr_df['Date of Termination'], errors = 'coerce')


production_df = pd.DataFrame(pg_list[1], columns = info_df.loc[info_df[2] == 'production_staff'][3].tolist(), \
                            index = np.array(pg_list[1])[:,0])
production_df = production_df.drop('id', axis=1)
production_df['Date of Hire'] = pd.to_datetime(production_df['Date of Hire'], errors = 'coerce')
production_df['TermDate'] = pd.to_datetime(production_df['TermDate'], errors = 'coerce')


recruit_df = pd.DataFrame(pg_list[2], columns = info_df.loc[info_df[2] == 'recruiting_costs'][3].tolist(), \
                         index = np.array(pg_list[2])[:,0])
recruit_df = recruit_df.drop('id', axis=1)


salary_df = pd.DataFrame(pg_list[3], columns = info_df.loc[info_df[2] == 'salary_grid'][3].tolist(), \
                        index = np.array(pg_list[3])[:,0])
salary_df = salary_df.drop('id', axis=1)

pd.options.display.max_columns = 40

Предположим, что hr-отделу нужно:
1. Исследовать параметр Days Employed среди уволенных сотрудников (для неуволенных мы не можем знать, сколько времени они проработают).
2. Увеличить Performance Score, набирая классы работников, которые ранее показали большое значение этого показателя.


Примем, что Performance Score у эффективных сотрудников = ['Exceeds', 'Exceptional', 'Fully Meets'], а у неэффективных = 'Needs Improvement', остальные непонятные характеристики - в отдельную группу Others.

1. marriedid и Days Employed

In [189]:
pd.DataFrame(pg_list[5], columns = columns_list[5])

marriedid  count of workers average Days Employed
0          0                57  764.7368421052631579
1          1                46  767.2173913043478261

В среднем работники с marriedid = 1 работают в компании чуть дольше, чем неженатые.

2. marriedid и Performance Score

In [282]:
pd.DataFrame(pg_list[6], columns = columns_list[6])

Performance  marriedid  quantity of workers per group  \
0     Good performance          0                            134   
1     Good performance          1                             84   
2  No good performance          0                              9   
3  No good performance          1                              6   
4               Others          0                             44   
5               Others          1                             33   

   workers in group marriedid  share of workers  
0                         187          0.716578  
1                         123          0.682927  
2                         187          0.048128  
3                         123          0.048780  
4                         187          0.235294  
5                         123          0.268293

Доля работников с marriedid = 0 с высокой производительностью = 0.7165778  
Доля работников с marriedid = 1 с высокой производительностью = 0.6829267  
По работникам с плохой производительностью данных мало - менее 10 случаев на группу

3. maritalstatusid и Days Employed

In [192]:
pd.DataFrame(pg_list[7], columns = columns_list[7])

maritalstatusid  count of workers  average Days Employed
0                0                36   879.1111111111111111
1                1                46   767.2173913043478261
2                3                 1  1675.0000000000000000
3                4                 4   414.5000000000000000
4                2                16   538.0625000000000000

Не корректно делать выводы по сотрудникам с maritalstatusid = 3 и 4, т.к. мало статистики собрано.
Из остальных предпочтительной группой является maritalstatusid = 0, следующая группа с maritalstatusid = 1 чуть хуже.

4. maritalstatusid и Performance Score

In [194]:
pd.DataFrame(pg_list[8], columns = columns_list[8])

Performance  maritalstatusid  quantity of workers per group  \
0      Good performance                0                             99   
1      Good performance                1                             84   
2      Good performance                2                             21   
3      Good performance                3                             10   
4      Good performance                4                              4   
5   No good performance                0                              7   
6   No good performance                1                              6   
7   No good performance                2                              2   
8                Others                0                             31   
9                Others                1                             33   
10               Others                2                              7   
11               Others                3                              2   
12               Others                4                              4   

    workers withmaritalstatusid  share of workers  
0                           137          0.722628  
1                           123          0.682927  
2                            30          0.700000  
3                            12          0.833333  
4                             8          0.500000  
5                           137          0.051095  
6                           123          0.048780  
7                            30          0.066667  
8                           137          0.226277  
9                           123          0.268293  
10                           30          0.233333  
11                           12          0.166667  
12                            8          0.500000

Доля работников с maritalstatusid = 0 с высокой производительностью = 0.722628  
Доля работников с maritalstatusid = 1 с высокой производительностью = 0.682927  
Доля работников с maritalstatusid = 2 с высокой производительностью = 0.7  
Доля работников с maritalstatusid = 3 с высокой производительностью = 0.833333  

Группу с maritalstatusid = 4 исключил, тк мало экспериментов. Лучшая оказалась группа maritalstatusid = 3, но лучше получить больше данных.

5. genderid и Days Employed

In [195]:
pd.DataFrame(pg_list[9], columns = columns_list[9])

genderid  count of workers average Days Employed
0         0                60  750.8000000000000000
1         1                43  786.8372093023255814

Работники с genderid = 0 (женщины) в среднем работают дольше, чем работники с genderid = 1 (мужчины).

6. genderid и Performance Score

In [196]:
pd.DataFrame(pg_list[10], columns = columns_list[10])

Performance  genderid  quantity of workers per group  \
0     Good performance         0                            122   
1     Good performance         1                             96   
2  No good performance         0                              5   
3  No good performance         1                             10   
4               Others         0                             50   
5               Others         1                             27   

   workers withgenderid  share of workers  
0                   177          0.689266  
1                   133          0.721805  
2                   177          0.028249  
3                   133          0.075188  
4                   177          0.282486  
5                   133          0.203008

Доля работников с genderid = 0 (женщины) с высокой производительностью = 0.6892655367231638  
Доля работников с genderid = 1 (мужчины) с высокой производительностью = 0.7218045112781954

7. empstatus_id и Days Employed

In [299]:
pd.DataFrame(pg_list[11], columns = columns_list[11])

empstatus_id  count of workers average Days Employed
0             1                 1   58.0000000000000000
1             5                88  781.9090909090909091
2             4                14  715.4285714285714286

Вошли не все значения, т.к. в выборку включены только сотрудники с Date of Termination, не равным нулю, т.е. работающие и отпускники не попали в таблицу.
В выборку почем

8. empstatus_id и Performance Score

In [199]:
pd.DataFrame(pg_list[12], columns = columns_list[12])

Performance  empstatus_id  quantity of workers per group  \
0      Good performance             1                            140   
1      Good performance             2                              2   
2      Good performance             3                             12   
3      Good performance             4                              6   
4      Good performance             5                             58   
5   No good performance             1                              7   
6   No good performance             4                              4   
7   No good performance             5                              4   
8                Others             1                             36   
9                Others             2                              9   
10               Others             3                              2   
11               Others             4                              4   
12               Others             5                             26   

    workers withempstatus_id  share of workers  
0                        183          0.765027  
1                         11          0.181818  
2                         14          0.857143  
3                         14          0.428571  
4                         88          0.659091  
5                        183          0.038251  
6                         14          0.285714  
7                         88          0.045455  
8                        183          0.196721  
9                         11          0.818182  
10                        14          0.142857  
11                        14          0.285714  
12                        88          0.295455

Не понял, что значит

Доля работников с empstatus_id = 1 с высокой производительностью = 0.765027  
Доля работников с empstatus_id = 1 с высокой производительностью = 0.181818  
Доля работников с empstatus_id = 4 с высокой производительностью = 0.6590909090909091

In [286]:
hr_df.head()

Employee Name  Employee Number  marriedid  maritalstatusid  \
1            Brown, Mia       1103024456          1                1   
2  LaRotonda, William         1106026572          0                2   
3      Steans, Tyrone         1302053333          0                0   
4       Howard, Estelle       1211050782          1                1   
5           Singh, Nan        1307059817          0                0   

   genderid  empstatus_id  deptid  perf_scoreid  age  Pay Rate state   zip  \
1         0             1       1             3   30     28.50    MA  1450   
2         1             1       1             3   34     23.00    MA  1460   
3         1             1       1             3   31     29.00    MA  2703   
4         0             1       1             9   32     21.50    MA  2170   
5         0             1       1             9   30     16.56    MA  2330   

         dob     sex maritaldesc citizendesc Hispanic/Latino  \
1 1987-11-24  Female     Married  US Citizen              No   
2 1984-04-26    Male    Divorced  US Citizen              No   
3 1986-09-01    Male      Single  US Citizen              No   
4 1985-09-16  Female     Married  US Citizen              No   
5 1988-05-19  Female      Single  US Citizen              No   

                    racedesc Date of Hire  Days Employed Date of Termination  \
1  Black or African American   2008-10-27           3317                 NaT   
2  Black or African American   2014-01-06           1420                 NaT   
3                      White   2014-09-29           1154                 NaT   
4                      White   2015-02-16             58          2015-04-15   
5                      White   2015-05-01            940                 NaT   

        Reason For Term Employment Status     department  \
1  N/A - still employed            Active  Admin Offices   
2  N/A - still employed            Active  Admin Offices   
3  N/A - still employed            Active  Admin Offices   
4  N/A - still employed            Active  Admin Offices   
5  N/A - still employed            Active  Admin Offices   

                   position        Manager Name         Employee Source  \
1              Accountant I  Brandon R. LeBlanc      Diversity Job Fair   
2              Accountant I  Brandon R. LeBlanc      Website Banner Ads   
3              Accountant I  Brandon R. LeBlanc         Internet Search   
4  Administrative Assistant  Brandon R. LeBlanc  Pay Per Click - Google   
5  Administrative Assistant  Brandon R. LeBlanc      Website Banner Ads   

          Performance Score  
1               Fully Meets  
2               Fully Meets  
3               Fully Meets  
4  N/A- too early to review  
5  N/A- too early to review

In [302]:
hr_df.query('empstatus_id == [1,5,4]')[['empstatus_id', 'Employment Status']].head(50)

empstatus_id       Employment Status
1              1                  Active
2              1                  Active
3              1                  Active
4              1                  Active
5              1                  Active
6              5  Voluntarily Terminated
7              1                  Active
8              5  Voluntarily Terminated
9              1                  Active
10             1                  Active
11             1                  Active
12             1                  Active
13             4    Terminated for Cause
14             4    Terminated for Cause
15             4    Terminated for Cause
16             1                  Active
18             1                  Active
19             5  Voluntarily Terminated
20             1                  Active
21             1                  Active
22             1                  Active
23             5  Voluntarily Terminated
24             1                  Active
25             1                  Active
26             1                  Active
27             1                  Active
28             5  Voluntarily Terminated
29             1                  Active
30             1                  Active
31             1                  Active
32             1                  Active
33             1                  Active
34             1                  Active
35             1                  Active
36             1                  Active
37             1                  Active
38             1                  Active
39             1                  Active
40             1                  Active
41             1                  Active
42             5  Voluntarily Terminated
43             1                  Active
44             5  Voluntarily Terminated
46             4    Terminated for Cause
47             5  Voluntarily Terminated
51             1                  Active
52             1                  Active
53             1                  Active
54             1                  Active
55             5  Voluntarily Terminated

In [300]:
hr_df.query('empstatus_id == [0, 2,3]')[['empstatus_id', 'Employment Status']]

empstatus_id Employment Status
17              3  Leave of Absence
45              2      Future Start
48              3  Leave of Absence
49              3  Leave of Absence
50              2      Future Start
73              3  Leave of Absence
79              3  Leave of Absence
84              3  Leave of Absence
91              2      Future Start
95              2      Future Start
121             3  Leave of Absence
143             3  Leave of Absence
155             3  Leave of Absence
178             2      Future Start
181             3  Leave of Absence
197             2      Future Start
205             2      Future Start
221             3  Leave of Absence
222             3  Leave of Absence
226             2      Future Start
227             2      Future Start
233             3  Leave of Absence
236             2      Future Start
242             3  Leave of Absence
277             2      Future Start

In [293]:
hr_df[['Employment Status', 'empstatus_id']].head(50)

Employment Status  empstatus_id
1                   Active             1
2                   Active             1
3                   Active             1
4                   Active             1
5                   Active             1
6   Voluntarily Terminated             5
7                   Active             1
8   Voluntarily Terminated             5
9                   Active             1
10                  Active             1
11                  Active             1
12                  Active             1
13    Terminated for Cause             4
14    Terminated for Cause             4
15    Terminated for Cause             4
16                  Active             1
17        Leave of Absence             3
18                  Active             1
19  Voluntarily Terminated             5
20                  Active             1
21                  Active             1
22                  Active             1
23  Voluntarily Terminated             5
24                  Active             1
25                  Active             1
26                  Active             1
27                  Active             1
28  Voluntarily Terminated             5
29                  Active             1
30                  Active             1
31                  Active             1
32                  Active             1
33                  Active             1
34                  Active             1
35                  Active             1
36                  Active             1
37                  Active             1
38                  Active             1
39                  Active             1
40                  Active             1
41                  Active             1
42  Voluntarily Terminated             5
43                  Active             1
44  Voluntarily Terminated             5
45            Future Start             2
46    Terminated for Cause             4
47  Voluntarily Terminated             5
48        Leave of Absence             3
49        Leave of Absence             3
50            Future Start             2

In [290]:
hr_df.empstatus_id.unique()

array([1, 5, 4, 3, 2])

9. deptid и Days Employed

In [201]:
pd.DataFrame(pg_list[13], columns = columns_list[13])

deptid  count of workers average Days Employed
0       6                 4  935.2500000000000000
1       1                 3  808.0000000000000000
2       3                10  428.8000000000000000
3       5                83  793.2891566265060241
4       4                 3  862.0000000000000000

10. deptid и Performance Score

In [202]:
pd.DataFrame(pg_list[14], columns = columns_list[14])

Performance  deptid  quantity of workers per group  \
0      Good performance       1                              7   
1      Good performance       2                              1   
2      Good performance       3                             32   
3      Good performance       4                              7   
4      Good performance       5                            147   
5      Good performance       6                             24   
6   No good performance       3                              1   
7   No good performance       4                              1   
8   No good performance       5                             11   
9   No good performance       6                              2   
10               Others       1                              3   
11               Others       3                             17   
12               Others       4                              2   
13               Others       5                             50   
14               Others       6                              5   

    workers withdeptid  share of workers  
0                   10          0.700000  
1                    1          1.000000  
2                   50          0.640000  
3                   10          0.700000  
4                  208          0.706731  
5                   31          0.774194  
6                   50          0.020000  
7                   10          0.100000  
8                  208          0.052885  
9                   31          0.064516  
10                  10          0.300000  
11                  50          0.340000  
12                  10          0.200000  
13                 208          0.240385  
14                  31          0.161290

11. state и Days Employed

In [203]:
pd.DataFrame(pg_list[15], columns = columns_list[15])

state  count of workers  average Days Employed
0    PA                 1   899.0000000000000000
1    TN                 1  1334.0000000000000000
2    OH                 1   425.0000000000000000
3    MA                98   759.2755102040816327
4    VA                 1  1083.0000000000000000
5    CT                 1   732.0000000000000000

12. deptid и Performance Score

In [204]:
pd.DataFrame(pg_list[16], columns = columns_list[16])

Performance state  quantity of workers per group  \
0      Good performance    AZ                              1   
1      Good performance    CA                              1   
2      Good performance    CO                              1   
3      Good performance    CT                              5   
4      Good performance    FL                              1   
5      Good performance    GA                              1   
6      Good performance    ID                              1   
7      Good performance    KY                              1   
8      Good performance    MA                            190   
9      Good performance    ME                              1   
10     Good performance    MT                              1   
11     Good performance    ND                              1   
12     Good performance    NH                              1   
13     Good performance    NY                              1   
14     Good performance    OH                              1   
15     Good performance    OR                              1   
16     Good performance    PA                              1   
17     Good performance    RI                              1   
18     Good performance    TN                              1   
19     Good performance    TX                              2   
20     Good performance    UT                              1   
21     Good performance    VA                              1   
22     Good performance    VT                              2   
23  No good performance    AL                              1   
24  No good performance    MA                             14   
25               Others    CT                              1   
26               Others    IN                              1   
27               Others    MA                             71   
28               Others    NC                              1   
29               Others    NV                              1   
30               Others    TX                              1   
31               Others    WA                              1   

    workers withstate  share of workers  
0                   1          1.000000  
1                   1          1.000000  
2                   1          1.000000  
3                   6          0.833333  
4                   1          1.000000  
5                   1          1.000000  
6                   1          1.000000  
7                   1          1.000000  
8                 275          0.690909  
9                   1          1.000000  
10                  1          1.000000  
11                  1          1.000000  
12                  1          1.000000  
13                  1          1.000000  
14                  1          1.000000  
15                  1          1.000000  
16                  1          1.000000  
17                  1          1.000000  
18                  1          1.000000  
19                  3          0.666667  
20                  1          1.000000  
21                  1          1.000000  
22                  2          1.000000  
23                  1          1.000000  
24                275          0.050909  
25                  6          0.166667  
26                  1          1.000000  
27                275          0.258182  
28                  1          1.000000  
29                  1          1.000000  
30                  3          0.333333  
31                  1          1.000000

13. state и Days Employed

In [205]:
pd.DataFrame(pg_list[17], columns = columns_list[17])

sex  count of workers average Days Employed
0  Female                60  750.8000000000000000
1    Male                43  786.8372093023255814

14. deptid и Performance Score

In [206]:
pd.DataFrame(pg_list[18], columns = columns_list[18])

Performance     sex  quantity of workers per group  \
0     Good performance  Female                            122   
1     Good performance    Male                             96   
2  No good performance  Female                              5   
3  No good performance    Male                             10   
4               Others  Female                             50   
5               Others    Male                             27   

   workers withsex  share of workers  
0              177          0.689266  
1              133          0.721805  
2              177          0.028249  
3              133          0.075188  
4              177          0.282486  
5              133          0.203008

15. state и Days Employed

In [207]:
pd.DataFrame(pg_list[19], columns = columns_list[19])

citizendesc  count of workers average Days Employed
0  Eligible NonCitizen                 5  867.8000000000000000
1          Non-Citizen                 3  697.3333333333333333
2           US Citizen                95  762.6421052631578947

16. deptid и Performance Score

In [208]:
pd.DataFrame(pg_list[20], columns = columns_list[20])

Performance          citizendesc  quantity of workers per group  \
0     Good performance  Eligible NonCitizen                              7   
1     Good performance          Non-Citizen                              4   
2     Good performance           US Citizen                            207   
3  No good performance  Eligible NonCitizen                              1   
4  No good performance           US Citizen                             14   
5               Others  Eligible NonCitizen                              4   
6               Others           US Citizen                             73   

   workers withcitizendesc  share of workers  
0                       12          0.583333  
1                        4          1.000000  
2                      294          0.704082  
3                       12          0.083333  
4                      294          0.047619  
5                       12          0.333333  
6                      294          0.248299

17. state и Days Employed

In [209]:
pd.DataFrame(pg_list[21], columns = columns_list[21])

"Hispanic/Latino"  count of workers average Days Employed
0                No                94  772.7234042553191489
1               Yes                 9  694.0000000000000000

18. deptid и Performance Score

In [210]:
pd.DataFrame(pg_list[22], columns = columns_list[22])

Performance "Hispanic/Latino"  quantity of workers per group  \
0     Good performance                No                            200   
1     Good performance               Yes                             16   
2     Good performance                no                              1   
3     Good performance               yes                              1   
4  No good performance                No                             14   
5  No good performance               Yes                              1   
6               Others                No                             67   
7               Others               Yes                             10   

   workers with"Hispanic/Latino"  share of workers  
0                            281          0.711744  
1                             27          0.592593  
2                              1          1.000000  
3                              1          1.000000  
4                            281          0.049822  
5                             27          0.037037  
6                            281          0.238434  
7                             27          0.370370

19. state и Days Employed

In [211]:
pd.DataFrame(pg_list[23], columns = columns_list[23])

racedesc  count of workers  average Days Employed
0                   Hispanic                 1     2.0000000000000000
1          Two or more races                 7  1100.5714285714285714
2                      Asian                11   679.0000000000000000
3                      White                67   742.5373134328358209
4  Black or African American                17   821.0000000000000000

20. deptid и Performance Score

In [212]:
pd.DataFrame(pg_list[24], columns = columns_list[24])

Performance                          racedesc  \
0      Good performance  American Indian or Alaska Native   
1      Good performance                             Asian   
2      Good performance         Black or African American   
3      Good performance                          Hispanic   
4      Good performance                 Two or more races   
5      Good performance                             White   
6   No good performance                             Asian   
7   No good performance         Black or African American   
8   No good performance                          Hispanic   
9   No good performance                 Two or more races   
10  No good performance                             White   
11               Others                             Asian   
12               Others         Black or African American   
13               Others                          Hispanic   
14               Others                 Two or more races   
15               Others                             White   

    quantity of workers per group  workers withracedesc  share of workers  
0                               4                     4          1.000000  
1                              24                    34          0.705882  
2                              42                    57          0.736842  
3                               2                     4          0.500000  
4                              13                    18          0.722222  
5                             133                   193          0.689119  
6                               1                    34          0.029412  
7                               7                    57          0.122807  
8                               1                     4          0.250000  
9                               1                    18          0.055556  
10                              5                   193          0.025907  
11                              9                    34          0.264706  
12                              8                    57          0.140351  
13                              1                     4          0.250000  
14                              4                    18          0.222222  
15                             55                   193          0.284974

21. state и Days Employed

In [213]:
pd.DataFrame(pg_list[25], columns = columns_list[25])

department  count of workers average Days Employed
0         Admin Offices                 3  808.0000000000000000
1                 IT/IS                10  428.8000000000000000
2  Software Engineering                 3  862.0000000000000000
3                 Sales                 4  935.2500000000000000
4     Production                       83  793.2891566265060241

22. deptid и Performance Score

In [214]:
pd.DataFrame(pg_list[26], columns = columns_list[26])

Performance            department  quantity of workers per group  \
0      Good performance         Admin Offices                              7   
1      Good performance      Executive Office                              1   
2      Good performance                 IT/IS                             32   
3      Good performance     Production                                   147   
4      Good performance                 Sales                             24   
5      Good performance  Software Engineering                              7   
6   No good performance                 IT/IS                              1   
7   No good performance     Production                                    11   
8   No good performance                 Sales                              2   
9   No good performance  Software Engineering                              1   
10               Others         Admin Offices                              3   
11               Others                 IT/IS                             17   
12               Others     Production                                    50   
13               Others                 Sales                              5   
14               Others  Software Engineering                              2   

    workers withdepartment  share of workers  
0                       10          0.700000  
1                        1          1.000000  
2                       50          0.640000  
3                      208          0.706731  
4                       31          0.774194  
5                       10          0.700000  
6                       50          0.020000  
7                      208          0.052885  
8                       31          0.064516  
9                       10          0.100000  
10                      10          0.300000  
11                      50          0.340000  
12                     208          0.240385  
13                      31          0.161290  
14                      10          0.200000

23. state и Days Employed

In [215]:
pd.DataFrame(pg_list[27], columns = columns_list[27])

"Manager Name"  count of workers  average Days Employed
0       Michael Albert                 9   653.3333333333333333
1           Simon Roup                 8   345.7500000000000000
2           John Smith                 3   947.3333333333333333
3             Amy Dunn                13   781.3076923076923077
4       Brannon Miller                 6  1059.0000000000000000
5       Webster Butler                13   978.1538461538461538
6   Brandon R. LeBlanc                 2   394.0000000000000000
7           Janet King                 6  1075.8333333333333333
8         Elijiah Gray                 8   926.5000000000000000
9      Alex Sweetwater                 3   862.0000000000000000
10       David Stanley                 6   347.3333333333333333
11        Peter Monroe                 1   127.0000000000000000
12     Jennifer Zamora                 1  1395.0000000000000000
13       Kelley Spirea                 6  1096.3333333333333333
14      Kissy Sullivan                12   496.3333333333333333
15       Ketsia Liebig                 5   777.4000000000000000
16      Debra Houlihan                 1   899.0000000000000000

24. deptid и Performance Score

In [216]:
pd.DataFrame(pg_list[28], columns = columns_list[28])

Performance      "Manager Name"  quantity of workers per group  \
0      Good performance     Alex Sweetwater                              6   
1      Good performance            Amy Dunn                             14   
2      Good performance  Board of Directors                              2   
3      Good performance  Brandon R. LeBlanc                              4   
4      Good performance      Brannon Miller                             15   
5      Good performance    Brian Champaigne                              8   
6      Good performance       David Stanley                             16   
7      Good performance      Debra Houlihan                              2   
8      Good performance        Elijiah Gray                             15   
9      Good performance        Eric Dougall                              4   
10     Good performance          Janet King                             17   
11     Good performance     Jennifer Zamora                              6   
12     Good performance          John Smith                             11   
13     Good performance       Kelley Spirea                             18   
14     Good performance       Ketsia Liebig                             16   
15     Good performance      Kissy Sullivan                             16   
16     Good performance       Lynn Daneault                             10   
17     Good performance      Michael Albert                             11   
18     Good performance        Peter Monroe                              6   
19     Good performance          Simon Roup                              8   
20     Good performance      Webster Butler                             13   
21  No good performance     Alex Sweetwater                              1   
22  No good performance      Brannon Miller                              1   
23  No good performance      Debra Houlihan                              1   
24  No good performance        Elijiah Gray                              1   
25  No good performance          Janet King                              2   
26  No good performance     Jennifer Zamora                              1   
27  No good performance          John Smith                              1   
28  No good performance       Ketsia Liebig                              1   
29  No good performance      Kissy Sullivan                              1   
30  No good performance      Michael Albert                              3   
31  No good performance      Webster Butler                              2   
32               Others     Alex Sweetwater                              2   
33               Others            Amy Dunn                              7   
34               Others  Brandon R. LeBlanc                              3   
35               Others      Brannon Miller                              5   
36               Others       David Stanley                              5   
37               Others        Elijiah Gray                              6   
38               Others          John Smith                              2   
39               Others       Kelley Spirea                              4   
40               Others       Ketsia Liebig                              4   
41               Others      Kissy Sullivan                              5   
42               Others       Lynn Daneault                              3   
43               Others      Michael Albert                              8   
44               Others        Peter Monroe                              8   
45               Others          Simon Roup                              9   
46               Others      Webster Butler                              6   

    workers with"Manager Name"  share of workers  
0                            9          0.666667  
1                           21          0.666667  
2                            2          1.000000  
3                            7          0.571429  
4           

25. state и Days Employed

In [218]:
pd.DataFrame(pg_list[29], columns = columns_list[29])

"Employee Source"  count of workers  \
0               Company Intranet - Partner                 1   
1                     Professional Society                 3   
2                  On-line Web application                 1   
3                                Billboard                 5   
4                            Pay Per Click                 1   
5                                 MBTA ads                 4   
6                       Newspager/Magazine                 5   
7                          Internet Search                 2   
8        Search Engine - Google Bing Yahoo                15   
9                   Pay Per Click - Google                 4   
10                             Monster.com                11   
11                                   Other                 3   
12  Social Networks - Facebook Twitter etc                 8   
13                           Word of Mouth                 7   
14                       Employee Referral                 4   
15                               Glassdoor                 6   
16                         Vendor Referral                 4   
17                      Website Banner Ads                 1   
18                    On-campus Recruiting                 1   
19                     Information Session                 1   
20                      Diversity Job Fair                16   

    average Days Employed  
0    444.0000000000000000  
1   1202.3333333333333333  
2    194.0000000000000000  
3    577.8000000000000000  
4      2.0000000000000000  
5    989.0000000000000000  
6    482.6000000000000000  
7   1057.5000000000000000  
8    828.0666666666666667  
9    419.5000000000000000  
10  1020.4545454545454545  
11   858.3333333333333333  
12   715.6250000000000000  
13   400.5714285714285714  
14   867.5000000000000000  
15   787.5000000000000000  
16   469.7500000000000000  
17  1575.0000000000000000  
18    69.0000000000000000  
19  1271.0000000000000000  
20   865.3125000000000000

26. deptid и Performance Score

In [219]:
pd.DataFrame(pg_list[30], columns = columns_list[30])

Performance                       "Employee Source"  \
0      Good performance                               Billboard   
1      Good performance                           Careerbuilder   
2      Good performance              Company Intranet - Partner   
3      Good performance                      Diversity Job Fair   
4      Good performance                       Employee Referral   
5      Good performance                               Glassdoor   
6      Good performance                                  Indeed   
7      Good performance                     Information Session   
8      Good performance                         Internet Search   
9      Good performance                                MBTA ads   
10     Good performance                             Monster.com   
11     Good performance                      Newspager/Magazine   
12     Good performance                    On-campus Recruiting   
13     Good performance                 On-line Web application   
14     Good performance                                   Other   
15     Good performance                  Pay Per Click - Google   
16     Good performance                    Professional Society   
17     Good performance       Search Engine - Google Bing Yahoo   
18     Good performance  Social Networks - Facebook Twitter etc   
19     Good performance                         Vendor Referral   
20     Good performance                      Website Banner Ads   
21     Good performance                           Word of Mouth   
22  No good performance                               Billboard   
23  No good performance                      Diversity Job Fair   
24  No good performance                               Glassdoor   
25  No good performance                         Internet Search   
26  No good performance                                MBTA ads   
27  No good performance                             Monster.com   
28  No good performance                                   Other   
29  No good performance                  Pay Per Click - Google   
30  No good performance       Search Engine - Google Bing Yahoo   
31  No good performance                           Word of Mouth   
32               Others                               Billboard   
33               Others                      Diversity Job Fair   
34               Others                       Employee Referral   
35               Others                               Glassdoor   
36               Others                     Information Session   
37               Others                         Internet Search   
38               Others                                MBTA ads   
39               Others                             Monster.com   
40               Others                      Newspager/Magazine   
41               Others                    On-campus Recruiting   
42               Others                                   Other   
43               Others                           Pay Per Click   
44               Others                  Pay Per Click - Google   
45               Others                    Professional Society   
46               Others       Search Engine - Google Bing Yahoo   
47               Others  Social Networks - Facebook Twitter etc   
48               Others                         Vendor Referral   
49               Others                      Website Banner Ads   
50               Others                           Word of Mouth   

    quantity of workers per group  workers with"Employee Source"  \
0                              12                             16   
1                               1                              1   
2                               1                              1   
3                              20                             29   
4                              21                             31   
5                              10                             14   
6                               8                  

In [265]:
pd.DataFrame(pg_list[31]).shape

(1, 1)

In [269]:
pd.DataFrame(pg_list[31], columns = [columns_list[31]])

count of maritalstatusid <> maritaldesc
0                                       0

In [279]:
pd.DataFrame(pg_list[32], columns = [columns_list[32]])

genderid <> sex
0               0

In [297]:
pd.DataFrame(pg_list[33], columns = [columns_list[33]])

empstatus_id <> Employment Status
0                                 0

In [23]:
hr_df.head()

Employee Name  Employee Number  marriedid  maritalstatusid  \
1            Brown, Mia       1103024456          1                1   
2  LaRotonda, William         1106026572          0                2   
3      Steans, Tyrone         1302053333          0                0   
4       Howard, Estelle       1211050782          1                1   
5           Singh, Nan        1307059817          0                0   

   genderid  empstatus_id  deptid  perf_scoreid  age  Pay Rate state   zip  \
1         0             1       1             3   30     28.50    MA  1450   
2         1             1       1             3   34     23.00    MA  1460   
3         1             1       1             3   31     29.00    MA  2703   
4         0             1       1             9   32     21.50    MA  2170   
5         0             1       1             9   30     16.56    MA  2330   

         dob     sex maritaldesc citizendesc Hispanic/Latino  \
1 1987-11-24  Female     Married  US Citizen              No   
2 1984-04-26    Male    Divorced  US Citizen              No   
3 1986-09-01    Male      Single  US Citizen              No   
4 1985-09-16  Female     Married  US Citizen              No   
5 1988-05-19  Female      Single  US Citizen              No   

                    racedesc Date of Hire  Days Employed Date of Termination  \
1  Black or African American   2008-10-27           3317                 NaT   
2  Black or African American   2014-01-06           1420                 NaT   
3                      White   2014-09-29           1154                 NaT   
4                      White   2015-02-16             58          2015-04-15   
5                      White   2015-05-01            940                 NaT   

        Reason For Term Employment Status     department  \
1  N/A - still employed            Active  Admin Offices   
2  N/A - still employed            Active  Admin Offices   
3  N/A - still employed            Active  Admin Offices   
4  N/A - still employed            Active  Admin Offices   
5  N/A - still employed            Active  Admin Offices   

                   position        Manager Name         Employee Source  \
1              Accountant I  Brandon R. LeBlanc      Diversity Job Fair   
2              Accountant I  Brandon R. LeBlanc      Website Banner Ads   
3              Accountant I  Brandon R. LeBlanc         Internet Search   
4  Administrative Assistant  Brandon R. LeBlanc  Pay Per Click - Google   
5  Administrative Assistant  Brandon R. LeBlanc      Website Banner Ads   

          Performance Score  
1               Fully Meets  
2               Fully Meets  
3               Fully Meets  
4  N/A- too early to review  
5  N/A- too early to review

In [280]:
production_df

Employee Name Race Desc Date of Hire   TermDate  \
1     Albert, Michael       White   2011-08-01        NaT   
2        Bozzi, Charles     Asian   2013-09-30 2014-08-07   
3    Butler, Webster  L     White   2016-01-28        NaT   
4           Dunn, Amy       White   2014-09-18        NaT   
5       Gray, Elijiah       White   2015-06-02        NaT   
..                  ...       ...          ...        ...   
252                None      None          NaT        NaT   
253                None      None          NaT        NaT   
254                None      None          NaT        NaT   
255                None      None          NaT        NaT   
256                None      None          NaT        NaT   

          Reason for Term       Employment Status         Department  \
1    N/A - still employed                  Active  Production          
2                retiring  Voluntarily Terminated  Production          
3    N/A - still employed                  Active  Production          
4    N/A - still employed                  Active  Production          
5    N/A - still employed                  Active  Production          
..                    ...                     ...                ...   
252                  None                    None               None   
253                  None                    None               None   
254                  None                    None               None   
255                  None                    None               None   
256                  None                    None               None   

               Position     Pay    Manager Name Performance Score  \
1    Production Manager  $54.50  Elisa Bramante       Fully Meets   
2    Production Manager  $50.50  Elisa Bramante       Fully Meets   
3    Production Manager  $55.00  Elisa Bramante           Exceeds   
4    Production Manager  $51.00  Elisa Bramante       Fully Meets   
5    Production Manager  $54.00  Elisa Bramante       Fully Meets   
..                  ...     ...             ...               ...   
252                None    None            None              None   
253                None    None            None              None   
254                None    None            None              None   
255                None    None            None              None   
256                None    None            None              None   

     Abutments/Hour Wk 1  Abutments/Hour Wk 2  Daily Error Rate  \
1                    0.0                  0.0               0.0   
2                    0.0                  0.0               0.0   
3                    0.0                  0.0               0.0   
4                    0.0                  0.0               0.0   
5                    0.0                  0.0               0.0   
..                   ...                  ...               ...   
252                  NaN                  NaN               NaN   
253                  NaN                  NaN               NaN   
254                  NaN                  NaN               NaN   
255                  NaN                  NaN               NaN   
256                  NaN                  NaN               NaN   

     90-day Complaints  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
5                  0.0  
..                 ...  
252                NaN  
253                NaN  
254                NaN  
255                NaN  
256                NaN  

[256 rows x 15 columns]

In [25]:
recruit_df

Employment Source  January   February  March  April  \
1                                Billboard       520       520    520    520   
2                            Careerbuilder       410       410    410    820   
3               Company Intranet - Partner         0         0      0      0   
4                       Diversity Job Fair         0      5129      0      0   
5                        Employee Referral         0         0      0      0   
6                                Glassdoor         0         0      0      0   
7                      Information Session         0         0      0      0   
8                          Internet Search         0         0      0      0   
9                                 MBTA ads       640       640    640    640   
10                             Monster.com       500       500    500    440   
11                      Newspager/Magazine       629       510    293    810   
12                    On-campus Recruiting         0         0   2500      0   
13                 On-line Web application         0         0      0      0   
14                                   Other         0       492      0    829   
15                           Pay Per Click       110       110     60    121   
16                  Pay Per Click - Google       330       330    180    362   
17                    Professional Society       100       100    100    100   
18       Search Engine - Google Bing Yahoo       330       410    388    372   
19  Social Networks - Facebook Twitter etc       420       481    452    479   
20                         Vendor Referral         0         0      0      0   
21                      Website Banner Ads       400       400    300    388   
22                           Word of Mouth         0         0      0      0   

    May  June  July  August  September  October  November  December  Total  
1     0     0   612     612        729      749       910       500   6192  
2   820   410   410     820        820     1230       820       410   7790  
3     0     0     0       0          0        0         0         0      0  
4     0     0     0    4892          0        0         0         0  10021  
5     0     0     0       0          0        0         0         0      0  
6     0     0     0       0          0        0         0         0      0  
7     0     0     0       0          0        0         0         0      0  
8     0     0     0       0          0        0         0         0      0  
9   640   640   640    1300       1300     1300      1300      1300  10980  
10  500   500   440     500        440      440       500       500   5760  
11  642   675   707     740        772      805       838       870   8291  
12    0  2500     0       0       2500        0         0         0   7500  
13    0     0     0       0          0        0         0         0      0  
14  744     0   610       0          0      510         0       810   3995  
15  110   109   130     146        105      109       105       110   1323  
16  197   152   389     437        315      327       315       176   3509  
17  100   100   100     100        100      100       100       100   1200  
18  472   412   416     495        619      502       389       378   5183  
19  392   508   578     466        389      439       491       478   5573  
20    0     0     0       0          0        0         0         0      0  
21  592   610   620     669        718      767       816       865   7143  
22    0     0     0       0          0        0         0         0      0

In [26]:
salary_df

Position  Salary Min  Salary Mid  Salary Max  \
1       Administrative Assistant       30000       40000       50000   
2   Sr. Administrative Assistant       35000       45000       55000   
3                   Accountant I       42274       51425       62299   
4                  Accountant II       50490       62158       74658   
5                 Sr. Accountant       63264       76988       92454   
6               Network Engineer       50845       66850       88279   
7           Sr. Network Engineer       79428       99458      120451   
8         Database Administrator       50569       68306       93312   
9                        Sr. DBA       92863      116007      139170   
10       Production Technician I       30000       40000       50000   
11      Production Technician II       38000       48000       58000   
12    Lead Production Technician       45000       55000       65000   

    Hourly Min  Hourly Mid  Hourly Max  
1        14.42       19.23       24.04  
2        16.83       21.63       26.44  
3        20.32       24.72       29.95  
4        24.27       29.88       35.89  
5        30.42       37.01       44.45  
6        24.44       32.14       42.44  
7        38.19       47.82       57.91  
8        24.31       32.84       44.86  
9        44.65       55.77       66.91  
10       14.42       19.23       24.04  
11       18.27       23.08       27.88  
12       21.63       26.44       31.25

1. hr_dataset_list имеет два поля 'sex' и 'genderid', которые хранят один и тот же параметр, но в разной форме.

## Часть вторая

Перед вами стоит задача – подготовить аналитический ответ для SMM-отдела компании Skillbox. <br> Объектом анализа является  [паблик Skillbox Вконтакте](https://vk.com/skillbox_education). <br> <br> 
Подключитесь к  API VK и выгрузите посты со стены паблика Skillbox за интересующий период (определите самостоятельно и обоснуйте). Проанализируйте влияние различных факторов (например, времени публикации) на вовлеченность пользователей (количество лайков, комментариев, голосов в опросах). Сделайте аналитику по рубрикам (примеры рубрик: дизайн-битва, игра по управлению), которые есть в паблике. Выбрать нужные посты можно с помощью регулярных выражений. Составьте перечень рекомандаций для SMM-отдела по итогам анализа. <br> <br> 

Дополнительные инструкции по работе с API VK расположены [здесь](https://nbviewer.jupyter.org/github/vikaborel/dstutorials/blob/master/Дополнительные%20инструкции.ipynb).

In [ ]:
### YOUR CODE HERE ###


